### PR TITLE
Protips - Part one

### PR DESCRIPTION
**Pull request by MaxIsJoe**
work on #5316 

Part one adds the following :

- [X] Add the UI and Manager
- [X] Add detection for players getting hit
- [X] Add detection for players picking up an item
- [X] Add detection for players being on low health
- [x] Add detection for players who are in range of an object
- [X] Improve UI animations
- [x] Add UI for changing player experience

Everything in this PR is a WIP and is meant to showcase the very bare minimum of this system and UI implemented to get things rolling.

Part two will add more triggers and a queue system for tips. 

CL: [New] added a basic foundation for the ProTips system for educating new players.

### CODE REVIEW

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 1]
```diff
@@ -248,6 +249,9 @@ public BleedingState CalculateBleedingState()
 
 		public PlayerScript playerScript;
 
+		public event Action<DamageType> OnTakeDamageType;
+		public UnityEvent OnLowHealth = new UnityEvent();
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs]: you have an example of c# event just above the one you did, why not vanilla c# events instead.

**Comment by corp-0** [comment]
if you make it so admins can create and trigger arbitrary protips that'd be incredibly memeable

**Comment by MaxIsJoe** [comment]
Review pls

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@
 using Player;
 using Newtonsoft.Json;
 using ScriptableObjects.RP;
+using Event = ENet.Event;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs]: this using statement isn't needed anymore and might bring confusion

File: UnityProject/Assets/Scripts/Learning/ProtipObjectTypes/ProtipObjectOnEnterRadius.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+					break;
+				}
+			}
+		}
+
+		[Serializable]
+		public struct ObjectCheckData
+		{
+			public GameObject GameObjectToCheck;
+			public ProtipSO AssoicateSo;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Learning/ProtipObjectTypes/ProtipObjectOnEnterRadius.cs]: typo

File: UnityProject/Assets/Scripts/Learning/ProtipObjectTypes/ProtipObjectOnReceiveDamageType.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Learning.ProtipObjectTypes
+{
+	public class ProtipObjectOnReceiveDamageType : ProtipObject
+	{
+		[SerializeField] private DamageType damageTypeThatTriggersTip;
+
+		public void Start()
+		{
+			PlayerManager.PlayerScript.playerHealth.OnTakeDamageType += DamageTypeSimilar;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Learning/ProtipObjectTypes/ProtipObjectOnReceiveDamageType.cs:12]: I think you also need to unsuscribe when the object is destroyed

File: UnityProject/Assets/Scripts/Systems/Inventory/Pickupable.cs [hunk 5]
```diff
@@ -42,6 +43,8 @@ public class Pickupable : NetworkBehaviour, IPredictedCheckedInteractable<HandAp
 
 	public ItemAttributesV2 ItemAttributesV2;
 
+	public readonly UnityEvent OnMoveToPlayerInventory = new UnityEvent();
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Systems/Inventory/Pickupable.cs]: use vanilla events

File: UnityProject/Assets/Scripts/Learning/ProtipSO.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+using System;
+using UnityEngine;
+
+namespace Learning
+{
+	[CreateAssetMenu(fileName = "Protip", menuName = "ScriptableObjects/Learning/ProtipSO")]
+    public class ProtipSO : ScriptableObject
+    {
+        public string TipTile;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Learning/ProtipSO.cs]: typo

File: UnityProject/Assets/Scripts/Systems/Inventory/Pickupable.cs [hunk 7]
```diff
@@ -215,6 +218,7 @@ private IEnumerator ServerPerformInteractionLogic(HandApply interaction)
 
 	private void PickupAnim(GameObject interactor)
 	{
+		OnMoveToPlayerInventory.Invoke();
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Systems/Inventory/Pickupable.cs]:
```suggestion
		OnMoveToPlayerInventory?.Invoke();
```

**Comment by corp-0** [comment]
does it have compiling errors?

**Comment by MaxIsJoe** [comment]
> does it have compiling errors?

No, in fact i was able to enter playmode just fine

**Comment by corp-0** [comment]
```
HelpWindow (UnityEngine.GameObject) needs to be in the allSpawnablePrefabs list and has been added. Since the list has been updated you NEED to commit the changed NetworkManager Prefab file
ProTipManager (UnityEngine.GameObject) needs to be in the allSpawnablePrefabs list and has been added. Since the list has been updated you NEED to commit the changed NetworkManager Prefab file
ProTipUI (UnityEngine.GameObject) needs to be in the allSpawnablePrefabs list and has been added. Since the list has been updated you NEED to commit the changed NetworkManager Prefab fileat Tests.Asset.NetworkManagerTests.SpawnableListTest () [0x00254] in /github/workspace/UnityProject/Assets/Tests/Asset/NetworkManagerTests.cs:82ScriptableObjects.Atmospherics.GasesSingleton SO took 51 ms to find! Try to serialize a reference to this SO singleton instead!
```

Also take a look at the codacy issues, it is just 4 of them

**Comment by MaxIsJoe** [comment]
> ```
> HelpWindow (UnityEngine.GameObject) needs to be in the allSpawnablePrefabs list and has been added. Since the list has been updated you NEED to commit the changed NetworkManager Prefab file
> ProTipManager (UnityEngine.GameObject) needs to be in the allSpawnablePrefabs list and has been added. Since the list has been updated you NEED to commit the changed NetworkManager Prefab file
> ProTipUI (UnityEngine.GameObject) needs to be in the allSpawnablePrefabs list and has been added. Since the list has been updated you NEED to commit the changed NetworkManager Prefab fileat Tests.Asset.NetworkManagerTests.SpawnableListTest () [0x00254] in /github/workspace/UnityProject/Assets/Tests/Asset/NetworkManagerTests.cs:82ScriptableObjects.Atmospherics.GasesSingleton SO took 51 ms to find! Try to serialize a reference to this SO singleton instead!
> ```
> 
> Also take a look at the codacy issues, it is just 4 of them

thats stupid, why does the network manager need to know about something that wont be even spawned by anything network/admin related?

File: UnityProject/Assets/Scripts/Learning/UI/ProtipUI.cs [hunk 8]
```diff
@@ -36,14 +36,14 @@ public void ShowTip(string tip, float showDuration = 25f, Sprite img = null, Spr
...
-				case SpriteAnimation.NONE:
-					break;
 				case SpriteAnimation.ROCKING:
 					StartCoroutine(DoImageRockAnimations());
 					break;
 				case SpriteAnimation.UPDOWN:
 					StartCoroutine(DoImageJumpAnimations());
 					break;
+				default:
+					break;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Learning/UI/ProtipUI.cs:46]: Log.Error the unexpected value you received. It has helped me before to find issues where I wouldn't think was one

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Learning/UI/ProtipUI.cs:46]: I think this should replace `NONE` so a log is not needed

**Comment by MaxIsJoe** [comment]
why does tests keep failing here

**Review by corp-0 — APPROVED** [review]

**Pull request merged**